org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java

```diff
@@ -12,92 +12,161 @@
 package org.jacoco.core.test.validation;
 
 import java.io.BufferedReader;
+import java.io.FileInputStream;
 import java.io.IOException;
+import java.io.InputStreamReader;
 import java.io.Reader;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.List;
-import java.util.Map;
-import java.util.NoSuchElementException;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import org.jacoco.core.analysis.IBundleCoverage;
+import org.jacoco.core.analysis.IClassCoverage;
+import org.jacoco.core.analysis.ICoverageNode;
+import org.jacoco.core.analysis.ILine;
+import org.jacoco.core.analysis.IPackageCoverage;
+import org.jacoco.core.analysis.ISourceFileCoverage;
+
 /**
- * Reads a single source file and allows access to it through special probe
- * comments in the following format <code>//$line-<i>tag</i>$.
+ * Reads a single source file and allows access to its line coverage. Lines may
+ * contains executable comments in the format
+ * 
+ * <pre>
+ * // > statement1(); statement2();
+ * </pre>
  */
 public class Source {
 
-	private static final Pattern TAG_PATTERN = Pattern
-			.compile("\\$line-(.*)\\$");
+	private static final String SRC_LOCATION = "src/";
+
+	private static final String SRC_ENCODING = "UTF-8";
+
+	private static final Pattern EXEC_COMMENT_PATTERN = Pattern
+			.compile("//\\s*>(.*)");
```

Comment:
I like omitting `>` as it is less verbose - but way harder for the parser: We have `//` e.g. in multi-line comments referencing a URL. Our current naive approach matches single lines only. We cannot detect whether we're within a multi-line comment.

Any ideas?

Reply:
@marchof of course you right that in general case real lexer is required, and I'm ashamed that missed/forgot this fact. And in general case lexer of source language - multiline string literals in Kotlin/Groovy, etc.

However in our case seems that the only problem is URLs. What do you think about regex with negative lookbehind to exclude them - `(?<!https?:)//(.*)` ?

My quick attempt

```diff
--- org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java
+++ org.jacoco.core.test/src/org/jacoco/core/test/validation/Source.java
@@ -45,7 +45,7 @@
 	private static final String SRC_ENCODING = "UTF-8";
 
 	private static final Pattern EXEC_COMMENT_PATTERN = Pattern
-			.compile("//\\s*>(.*)");
+			.compile("(?<!https?:)//(.*)");
 
 	/**
 	 * Represents a single line in a source file.
--- org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java
+++ org.jacoco.core.test/src/org/jacoco/core/test/validation/StatementParser.java
@@ -73,6 +73,7 @@
 		tokenizer = new StreamTokenizer(new StringReader(source));
 		tokenizer.resetSyntax();
 		tokenizer.whitespaceChars(' ', ' ');
+		tokenizer.whitespaceChars('>', '>');
 		tokenizer.whitespaceChars('\t', '\t');
 		tokenizer.wordChars('a', 'z');
 		tokenizer.wordChars('A', 'Z');
```

and few conversions of comments from `//` into `/*`, shows that this is sufficient.

And so far I can imagine only one highly unlikely corner-case when this can miss executable comment

```
	enum E {
		http
	}

	static void example(E e) {
		switch (e) {
		case http:// executable comment
		}
	}
```

Reply:
Ok, I'll give it a try.

Reply:
✔️ Done. Actually identified 2 instances, where I forgot `>`. So less errorprone now.

Reply:
which means that I also overlooked them 😉 Now I see - one is `assertFinally` in `Exceptions` and still wondering where is another one? 😆 So definitely less error prone 👍